### ISSUE TITLE
logging and error management

### ISSUE THREAD

**Issue by Geal** [streamnative/pulsar-rs#14]
currently it's a bit annoying to understand failures, since most of them will result in a `Error::Disconnected` without any context.
I propose two things:
- adding the log crate and integrating logs here and there
- error structures with more context, like which consumer id generated it, which topics. I'm not fond of using crates like error-chain or failure, it's easy enough to make the error structures manually

**Comment by stearnsc** [comment]
That sounds great.
`log` seems like a fine choice for a logging library.
The reason `Disconnected` is so prevalent was that it was difficult to maintain separation of concerns between the use-agnostic `Connection` (where errors almost always occur) and the `Consumer`s and/or `Producer`s using that connection (which have appropriate context). I'm weary of weakening that separation, however having the consumer and producer return a wrapped error with more context seems totally fine. Was that what you were thinking?

Re: using `failure`, at the time it was intended to be blessed future of rust error handling, but since they've pivoted to reworking the `Error` trait itself I'm far less tied to it. Building errors by hand seems fine.

**Comment by Geal** [comment]
yes, I'd like to have error enums for consumer and producer, and they would have a branch to contain a connection error.
I'm wondering about the `SharedError` type. I understand keeping around the error in a connection to check if the connection is still valid, but why store its members in a `Arc`? `SharedError` and `Connection` are not `Sync`, they do not look designed to be used from multiple threads?

**Comment by Geal** [comment]
oh, right, now I see, `TaskExecutor` requires a `Future<(), ()>`, so we need to store the error coming from the sender and receiver tasks. Is there a guarantee that there will only be one error in flight at any moment?

**Comment by stearnsc** [comment]
No there's not, but since each error is fatal, I decided it would be okay to only keep the first one. That said, I consider that a pretty hacky solution, and am open to a more elegant one :)

**Comment by Geal** [comment]
Since the protocol is multiplexed, allowing multiple producers and
consumers over the same connection, errors should be dispatched instead of
keeping only one. But that will only be useful when implementing
multiplexing :)
On that topic, I can probably reproduce the architecture I had in the lapin
AMQP client

**Comment by stearnsc** [comment]
Could you point to an example of how lapin does it? The specific issue I never found a good solution to was:
In order to send to all clients, either I have to send by reference (which makes for an awkward API on the clients side), or have an error that's `Clone`, which means I can't always keep original errors from other libs. I think with your new error handling we could obfuscate the reference thing and make ConnectionError variants own an `Arc<Error>`; was that what you were thinking, or did you come up with something better?

**Comment by Geal** [comment]
so, I was mistaken, I thought I had handled errors in lapin with `From` conversion here and there, then using `from_err()`, but apparently I did not.
I experimented a bit with it in those examples: https://github.com/wyyerd/pulsar-rs/issues/1#issuecomment-486674655

I'm not sure sending errors by reference will be a good idea (there's no guarantee they will be `Send`). I'd rather have the errors converted to the library's internal error management right away.
